### PR TITLE
Multipackge support

### PR DESCRIPTION
**Pull request by jaymzh**
Allow the `package` provider to take an array of packages to handle in one
transaction.

This solves two large problems:
* There are times when you cannot install two packages in sequence, like when a
  binary is moving between two packages - they *must* be done in the same
  transaction.
* When using Chef to install the vast majority of your base system, it
  can make imaging take a very, very long time because executing yum or apt once
  for every single package is painfully slow.

This solves both. The scaffolding is all there in the Package HWRP, plus the
underlying implementation for both apt and yum, the two I have access to test.

### CODE REVIEW

**Comment by lamont-granquist** [comment]
I'm back on monday and will review (although there's a ton of start-of-year meeting chaos all week...)

File: lib/chef/provider/package/yum.rb [hunk 1]
```diff
@@ -1129,15 +1149,43 @@ def install_package(name, version)
...
                 end
               end
+              # methods don't count for packages we won't be touching
+              next if RPMVersion.parse(current_version_array[idx]) == RPMVersion.parse(v)
+              methods << method
+            end
+            # We could split this up into two commands if we wanted to, but
+            # for now, just don't support this.
+            if methods.uniq.length > 1
+              raise Chef::Exceptions::Package, "Multipackage rule #{name} has a mix of upgrade and downgrade packages. Cannot proceed."
```
Comment (jaymzh) [inline lib/chef/provider/package/yum.rb]: downgrade is a special case though. Chef doesn't even support it in apt/deb, and only supports it with special exceptions in yum. yum won't upgrade and downgrade in one command, so I felt being a bit more conservative here was the right choice for now. If people feel strongly I can make it run two commands.

File: lib/chef/provider/package/yum.rb [hunk 2]
```diff
@@ -1064,6 +1068,9 @@ def load_current_resource
             parse_arch
           end
 
+          # normalize internal representation as an array
+          @new_resource.package_name(as_array(@new_resource.package_name))
```
Comment (jaymzh) [inline lib/chef/provider/package/yum.rb]: as_array is a generalized helper so I didn't have to add a helper for every possible thing. In this case, `package_name_array` operates on `new_resource`, but we need `@new_resource`, it's not clear to me these are the same and this is the only instance like that so I didn't make yet another helper/

File: lib/chef/provider/package/yum.rb [hunk 3]
```diff
@@ -1090,10 +1097,18 @@ def load_current_resource
...
-          @current_resource.version(installed_version)
-
-          @candidate_version = @yum.candidate_version(@new_resource.package_name, arch)
+          installed_version = []
+          @candidate_version = []
+          package_name_array.each do |pkg|
+            installed_version << @yum.installed_version(pkg, arch)
+            @candidate_version << @yum.candidate_version(pkg, arch)
+          end
+          #if installed_version.size == 1
```
Comment (jaymzh) [inline lib/chef/provider/package/yum.rb]: the commented out code? Yes, absolutely, this was leftover from the conversion. It's unclear that keeping package_name as an array always is the right call here (as opposed to always using the `*_array` helpers)... especially because we cannot do it for package_name on the new resource...

**Comment by jaymzh** [comment]
Using internal representations of arrays always is causing a lot of problems because of inconsistencies... it's easier to convert to arrays as needed, I"m going to back that out...

**Comment by jaymzh** [comment]
I cannot test resource reporting because I don't have a server that supports it... I would love it if someone who has access to one can try merginging my diff locally and trying it.

**Comment by jaymzh** [comment]
Found another bug while testing notifications... another diff coming.

**Comment by jaymzh** [comment]
OK this all works surprisingly well.

You can only notify if you give it a nice name - but that's not surprising. In the cases I want to do `package [crazy list of packages] do` I can't imagine wanting to notify it, and if I do, I must give it a nice name and then use `package_name` for my list of packages. We can document this as well.

**Comment by jaymzh** [comment]
Hmmm. A multipackage resource takes more time than all the individual resources... gotta figure out why. Probably a problem with the yumcache state.

**Comment by jaymzh** [comment]
Alright... this performs as well or better than current master.

**Comment by lamont-granquist** [comment]
well, yr gonna need some moar specs...

the yum and apt providers are looking even more messy, i think i may want to have a look at cleaning them up after you're done here.

**Comment by jaymzh** [comment]
I'll work on some specs, but I'd like to get any other concerns with the code in the meantime.

**Comment by jaymzh** [comment]
There's specs for package. I'll do specs for yum and apt soon.

**Comment by jaymzh** [comment]
I don't actually see failures in the appveyor thing... but either way around I'll add a few final tests probably tomorrow. In the meantime @lamont-granquist  review, please?

**Comment by someara** [comment]
I'm very much -1 on on "package %w(crazy list of packages) do". 

Notifications and subscriptions are fundamental to a promise system and this makes it weird.

package 'title' do
  package_name %(crazy ray)
end 

... fine, whatever.

Also, there needs to be support for this in ChefSpec on day 0.

Does this mean we can do yum groupinstall now?

**Comment by jaymzh** [comment]
Added more yum tests. @lamont-granquist  still waiting on you.

I'm not exactly sure how to prevent `name` from being an array while still allowing `package_name` to be one.

**Comment by jaymzh** [comment]
BTW, I"ve been running this on a subset of machines at Facebook for a few weeks now with success. Cut four minutes of our first chef run just be collapsing 73 consecutive packages into one resource.

**Comment by lamont-granquist** [comment]
So this is a bit confusing:

```
  * apt_package[["git", "dmidecode", "ltrace", "bc", "zsh", "vim", "curl", "strace", "lsof", "systemtap", "telnet", "mtr", "subversion", "nmap", "gdb", "traceroute", "tcpdump", "htop", "iptraf", "tmux", "nuttcp", "s3cmd", "source-highlight", "sysbench", "python3-pygments", "heirloom-mailx", "ack-grep"]] action install
    - install version ["1:1.9.1-1ubuntu0.1", "2.12-2", "0.7.3-4ubuntu5.1", "1.06.95-8ubuntu1", "5.0.2-3ubuntu6", "2:7.4.052-1ubuntu3", "7.35.0-1ubuntu2.3", "4.8-1ubuntu5", "4.86+dfsg-1ubuntu2", "2.3-1ubuntu1", "0.17-36build2", "0.85-2", "1.8.8-1ubuntu3.1", "6.40-0.2ubuntu1", "7.7.1-0ubuntu5~14.04.2", "1:2.0.20-0ubuntu0.1", "4.5.1-2ubuntu1.1", "1.0.2-3", "3.0.0-8.1", "1.8-5", "6.1.2-4", "1.1.0~beta3-2", "3.1.6-2ubuntu1", "0.4.12-1.1", "1.6+dfsg-1ubuntu1", "12.5-2", "2.12-1"] of package ["git", "dmidecode", "ltrace", "bc", "zsh", "vim", "curl", "strace", "lsof", "systemtap", "telnet", "mtr", "subversion", "nmap", "gdb", "traceroute", "tcpdump", "htop", "iptraf", "tmux", "nuttcp", "s3cmd", "source-highlight", "sysbench", "python3-pygments", "heirloom-mailx", "ack-grep"]
```

All i did there was remove dmidecode and bc and its real difficult to tell what its actually installing.

File: lib/chef/provider/package.rb [hunk 4]
```diff
@@ -58,10 +74,10 @@ def define_resource_requirements
 
       def action_install
         # If we specified a version, and it's not the current version, move to the specified version
-        if !@new_resource.version.nil? && !(target_version_already_installed?)
+        if new_version_array.any? && !(target_version_already_installed?)
           install_version = @new_resource.version
         # If it's not installed at all, install it
-        elsif @current_resource.version.nil?
+        elsif current_version_array.any? { |x| x.nil? }
```
Comment (lamont-granquist) [inline lib/chef/provider/package.rb]: `include?(nil)` here?

File: lib/chef/provider/package.rb [hunk 5]
```diff
@@ -84,17 +100,18 @@ def action_install
...
-        else
-          @new_resource.version(candidate_version)
-          orig_version = @current_resource.version || "uninstalled"
-          converge_by("upgrade package #{@new_resource.package_name} from #{orig_version} to #{candidate_version}") do
-            upgrade_package(@new_resource.package_name, candidate_version)
-            Chef::Log.info("#{@new_resource} upgraded from #{orig_version} to #{candidate_version}")
-          end
+          return
+        end
+        @new_resource.version(candidate_version)
```
Comment (lamont-granquist) [inline lib/chef/provider/package.rb]: this mutates the new_resource which is a bug in the existing implementation.

File: lib/chef/provider/package.rb [hunk 6]
```diff
@@ -84,17 +100,18 @@ def action_install
...
+        orig_version = @current_resource.version || "uninstalled"
```
Comment (lamont-granquist) [inline lib/chef/provider/package.rb]: this should probably coerce `@current_resource.version` into an array and then replace all nil/false with "uninstalled"

File: lib/chef/provider/package.rb [hunk 7]
```diff
@@ -84,17 +100,18 @@ def action_install
...
+        converge_by("upgrade package #{@new_resource.package_name} from #{orig_version} to #{candidate_version}") do
+          upgrade_package(@new_resource.package_name, candidate_version)
+          Chef::Log.info("#{@new_resource} upgraded from #{orig_version} to #{candidate_version}")
```
Comment (lamont-granquist) [inline lib/chef/provider/package.rb]: You probably need array-aware multipackage helpers here for output.  The whole converge_by text should get rendered by a helper.  There also need to be helpers to diff the current versions from the candidate versions and array subtract off what doesn't change.

**Comment by lamont-granquist** [comment]
@jaymzh i'm working on an incoming patch for you, i spotted a few more deeper problems in the base package provider

**Comment by jaymzh** [comment]
For your apt-thing, output formatters are pretty limited in what I can do... the normal output is a bit more specific, but also... somehat hard to read because I didn't want to special-case it.

File: lib/chef/provider/package.rb (hunk 5 above)
Comment (jaymzh) [inline lib/chef/provider/package.rb]: yeah... but I'm trying to keep this as close to the original as possible and then you can make further fixes. I want to get this merged

File: lib/chef/provider/package.rb (hunk 6 above)
Comment (jaymzh) [inline lib/chef/provider/package.rb]: can do.

File: lib/chef/provider/package.rb (hunk 7 above)
Comment (jaymzh) [inline lib/chef/provider/package.rb]: ok. I sorta of avoided that because your initial comments seemed to want no special casing of multipackages, but I'm happy to do that.

File: lib/chef/provider/package.rb (hunk 4 above)
Comment (jaymzh) [inline lib/chef/provider/package.rb]: oh, duh. :) I'll change that.

**Comment by lamont-granquist** [comment]
I've got a bunch of work I'm doing on top of this here: https://github.com/chef/chef/tree/lcg/2692

it fixes define_resource_requirements to work, and fixes it to be more idempotent and fixes the output to be more informative about what was actually done.  Still tracking down some spec test failures, and I need to add more specs around installing and upgrading.

**Comment by lamont-granquist** [comment]
I think I'm missing some specs, but I think I'm reasonably happy with 26933eaf3f653da03612a5df5a2b47629845c24f if you want to pull that into here.

**Comment by jaymzh** [comment]
Made a few comments over there, dunno if you get notified since it's not in a PR... after that I'll pull it in to this. Really nice work.

**Comment by lamont-granquist** [comment]
Added 894cb01a38232b9006773dc27e053c9bbb4b5038 on top of that

I'm not sure what you want to do with that one bit of tortured logic, feel free to patch it after you pull it in.

Also note that I rebased your commits onto friday's master.  You might want to reset your head to 894cb01a38232b9006773dc27e053c9bbb4b5038 to pick up the rebase or just cherry pick those two.  Or whatever...

**Comment by jaymzh** [comment]
Cherry-picked your two fixes, swapped the order of that one conditional and updated. I'll do some testing at the office tomorrow, but I think we're in good shape to merge... I hope. :)

**Comment by lamont-granquist** [comment]
Yeah, lemme know if there's any perf hit, I wasn't being very conscious of that when I wrote it.

File: lib/chef/provider/package/apt.rb [hunk 8]
```diff
@@ -51,54 +51,87 @@ def default_release_options
...
-                # Returns all lines after 'Reverse Provides:'
-                showpkg.rpartition(/Reverse Provides:\s*#{$/}/)[2].each_line do |line|
-                  provider, version = line.split
-                  providers[provider] = version
+          final_installed_version = []
+          final_candidate_version = []
+          final_installed = []
+          final_virtual = []
+          installed = virtual = false
+          installed_version = candidate_version = nil
```
Comment (lamont-granquist) [inline lib/chef/provider/package/apt.rb]: can those two lines get dropped?

**Comment by lamont-granquist** [comment]
Okay, there's code cleanup I'd like to do in apt+yum providers, but it goes beyond the scope of this patch, I'm satisfied with the state of this yak.

**Comment by lamont-granquist** [comment]
@chef/client-core @chef/client-engineers final review before merge?

**Comment by jaymzh** [comment]
Testing looks good. Nuked those two lines. Ready when you are :)

**Comment by jaymzh** [comment]
Fucking appveyor. It seems to fail spuriously... and in fact this failure is almost certainly unrelated...

**Comment by jaymzh** [comment]
Crap. My local tests were invalid. Too many tests on tests. Let me validate the recent changes more thoroughly.

**Comment by jaymzh** [comment]
I found a bug... "provides" mapping isn't working properly.

```
package 'perl-DateManip' do
  action :upgrade
end
```

works because it maps it to 'perl-Date-Manip', but if you have that in a list, it does not work. I'll dig into that.

**Comment by jaymzh** [comment]
Bug fixed!

**Comment by jaymzh** [comment]
@lamont-granquist  wtf is `undefined method `orig_run' for class `Class'` ?

**Comment by lamont-granquist** [comment]
rebase on master maybe its fixed?

**Comment by danielsdeleo** [comment]
tl;dr, we need to lock RSpec to 3.1.x for the moment until we can update audit mode to work with newer RSpec.

@jaymzh I think this is related to RSpec 3.2.0 being released today. In order to use rspec to test the audit mode feature (which is itself using rspec) you have to sandbox rspec to not interfere with itself. Looks like we had to use non-public APIs for this, which got changed as part of making the rspec sandbox stuff a public API.

**Comment by jaymzh** [comment]
rebased. Had to force-push to my branch... something had gone wrong, but this looks right to me. Lets see how things go.

note that the most recent thing I pulled in from upstream/master is 5 days ago...

**Comment by jaymzh** [comment]
same rspec failures.

**Comment by lamont-granquist** [comment]
They can be ignored for this PR, like dan said master broke because rspec 3.2.0 got released

**Comment by jaymzh** [comment]
Sweet. I'll let you merge when ready. I added docs just now.

**Comment by tyler-ball** [comment]
Your changes look :+1: to me

**Comment by jaymzh** [comment]
Re-pushing with rebase

**Comment by jaymzh** [comment]
FTR, @lamont-granquist  - while your rspec change fixed the audit stuff, it broke the rubygems provider tests :)

**Comment by lamont-granquist** [comment]
Can you cherry-pick 0eb701a51896c53022c748200c905d3d8232e0a3 and see if that fixes it?

**Comment by jaymzh** [comment]
No sir:

```
         expected: ("gem install rspec-core -q --no-rdoc --no-ri -v \"\" -i /alt/install/location", {:env=>nil})
              got: ("gem install \"rspec-core\" -q --no-rdoc --no-ri  -i /alt/install/location", {:env=>nil})
```

but I can probably fix that for ya

**Comment by lamont-granquist** [comment]
(╯°□°）╯︵ ┻━┻

**Comment by jaymzh** [comment]
wtf, and now it's failing with the command it got last time... lemme revert that?

**Comment by jaymzh** [comment]
OK, so LOCALLY I get:

```
gem install \"rspec-core\" -q --no-rdoc --no-ri  -i /alt/install/location
```

but travis gets:

```
gem install rspec-core -q --no-rdoc --no-ri -v \"3.2.0\" -i /alt/install/location
```

wtf.

**Comment by jaymzh** [comment]
I give up. I have no idea how to fix that test

**Comment by jaymzh** [comment]
any ideas @lamont-granquist  ?

**Comment by lamont-granquist** [comment]
caffeinating and waking up...

**Comment by lamont-granquist** [comment]
try:

``
            expected ="gem install rspec-core -q --no-rdoc --no-ri -v \"#{@provider.candidate_version}\" -i /alt/install/location"
``

**Comment by jaymzh** [comment]
@lamont-granquist  as I said... I get a different result than travis gets... so hard to say.

On local box, that gets me:

```
         expected: ("gem install rspec-core -q --no-rdoc --no-ri -v \"\" -i /alt/install/location", {:env=>nil})
              got: ("gem install \"rspec-core\" -q --no-rdoc --no-ri  -i /alt/install/location", {:env=>nil})
```

but I can push that to travis and see if it works...

**Comment by jaymzh** [comment]
OK, that looks good. I don't understand why that fails for me with rvm on ruby 2.0.0, but whatevs. We're passing and ready to merge! :)

**Comment by lamont-granquist** [comment]
md files are there, so whacking merge...

**Comment by jaymzh** [comment]
I just found a bug in one of the error messages we can throw - we filter down the list of packages, but not the list of installed versions, so in the event of a "you're requesting an older package", we end up saying "[pkg1, pkg2] version [v1, v2, v3, v4, v5, v6 .... ] is newer than..." which is super not useful. I'll send a follow-up PR shortly.

**Comment by lamont-granquist** [comment]
So one thought about this PR is that we could fix #to_s on Chef::Resource to render names-as-arrays differently so that instead of this:

```
apt_package[["git", "dmidecode", "ltrace"]] action install (up to date)
```

We get:

```
apt_package[git, dmidecode, ltrace] action install (up to date)
```

That'd be a little bit more human-friendly for output and for notifies/subscribes.

**Comment by jaymzh** [comment]
That's a good idea... but I'm out of the office for the next few weeks, so I won't get to it soon.

Also, someone should actually try this with reporting... I don't have reporting setup on any server.

**Comment by lamont-granquist** [comment]
Tried it on hosted and it POSTs fine, the name of the resource is rendered as a string rather than being an array in the JSON, and the server consumes it fine.

**Comment by jaymzh** [comment]
:+1:
Thanks!

**Comment by martinb3** [comment]
This appears to break support for `action :none` on a package resource. Any action other than `:install` or `:upgrade` appears to trigger an exception from https://github.com/chef/chef/blob/master/lib/chef/provider/package.rb#L369-L372.

**Comment by jaymzh** [comment]
Yeah, that's a bug. Came in the refactor I pulled in from Lamont. I can have a look when I get back but I'm heading out to a dr. appt. @lamont-granquist may be able to look sooner. Without digging into the code right now, it's unclear to me if we should just be short-circuiting somewhere and not calling that on `action :nothing` or if that function should return something meaningful(ish).
